feat: Generic traits

RustanLeino opened the pull request:
Traits and classes that extend traits can now take type parameters.

A class is not allowed to extend a trait with more than one set of type parameters. For example, a class is not allow to extend both `MyTrait<int>` and `MyTrait<bool>`.

Co-authored-by: @robin-aws

lavaleri commented on Source/Dafny/DafnyAst.cs:
```diff
@@ -9039,27 +9243,75 @@ public class MemberSelectExpr : Expression {
+      var subst = new Dictionary<TypeParameter, Type>();
+      var i = 0;
+      foreach (var tp in cl.TypeArgs) {
+        subst.Add(tp, TypeApplication[i]);
```
We can get an Index out of bounds if we have a class such as myClass\<T, R\> extends myTrait\<T\>. SImplest repro I can come up with:
```
module TraitBug {
    trait MyTrait<T> {
        ghost const Repr: set<object>;
    }

    class MyClass<T, R> extends MyTrait<T> {
        constructor() {
            this.Repr := {this};
        }
    }

    class OtherClass {
        ghost const Repr: set<object>;
        var foo: MyTrait<int>;

        constructor()
        {
            var foo := new MyClass<int, int>();
            this.foo := foo;
            this.Repr := {this} + foo.Repr; // This is where it runs into trouble
        }
    }

    method someMethod() {
        var foo := new OtherClass();
    }
}
```

RustanLeino replied:
I've distilled this example into `module ReceiverResolution` in TraitVerify.dfy.

lavaleri commented:
Documenting a current gap using this with the Go compiler:

Generic traits with Go do not work if a Dafny class which implements the trait does not also use a generic type. example:
```
module GoCompileError {

  trait MyTrait<T> {
    method DoSomething(param: T)
  }

  // ClassA is able to compile
  class ClassA<T> extends MyTrait<T> {
    method DoSomething(param: T) {}
  }

  // The compilation of ClassB is invalid
  class ClassB extends MyTrait<int> {
    method DoSomething(param: int) {}
  }
}
```

The is because the compiled stuct `ClassB` has a function with signature `DoSomething(int)`, but the compiled interface `MyTrait` expects a function with signature `DoSomething(interface{})`.

RustanLeino commented:
@lavaleri , I've fixed the Module1/Module2 problem in commit https://github.com/dafny-lang/dafny/pull/619/commits/0b282e35d431434c8b2e63c41df4c0c3d2256ca2 that is part of the later PR https://github.com/dafny-lang/dafny/pull/619 ("traits extending traits"). I added it as an additional test case in that PR (commit https://github.com/dafny-lang/dafny/pull/619/commits/0a06c8480e451ac5f58ed296d7dc22937a89e01e).
